### DISCUSSION TITLE
Make a field optional depending on the value of another attribute

### DISCUSSION THREAD

**Discussion by jeanlst** [pydantic/pydantic#3673] [Question] [answered]
Hey guys,

Is there any way to make a field optional depending on the value of another attribute?
```python
class MyEnum(enum.Enum):
    Manual = "manual"
    Automatic = "automatic"
class Teste(BaseModel):
   base: MyEnum
   top: Optional[str]
```
In here I would like to make `top` optional if `base = MyEnum.automatic` and required if `base = MyEnum.manual`

I have tried using a `validator` but when throwing a `ValueError` it doesn't work well with `FastAPI`, since it's gonna return an `Internal Server Error` instead of returning a `422 Unprocessable Entity` with stuff like `"msg": "field required", "type": "value_error.missing"`

**Comment by uriyyo** [comment, marked as answer]
`validator` should work for such case. Maybe you missed `always=True` when declatate the validator?
```py
import enum
from typing import Optional, Any, Mapping

from pydantic import BaseModel, validator
from pydantic.errors import NoneIsNotAllowedError


class MyEnum(enum.Enum):
    Manual = "manual"
    Automatic = "automatic"


class Teste(BaseModel):
    base: MyEnum
    top: Optional[str]

    @validator("top", always=True)
    def __validate_top__(cls, v: Optional[str], values: Mapping[str, Any]) -> Any:
        if values["base"] == MyEnum.Manual and v is None:
            raise NoneIsNotAllowedError

        return v


a = Teste(base=MyEnum.Automatic)  # no error
b = Teste(base=MyEnum.Manual, top="value")  # no error
c = Teste(base=MyEnum.Manual)  # will raise ValidationError
```
```py
pydantic.error_wrappers.ValidationError: 1 validation error for Teste
top
  none is not an allowed value (type=type_error.none.not_allowed)
```

**Reply by jeanlst** [reply]
Thank you @uriyyo , that did the trick. That's probably what happened. I was wondering, should I also have `pre=True` just in case?

**Reply by uriyyo** [reply]
I think it's better not to use `pre=True` in this case. When you use `pre=True` then the validator will be called before parsing value, but in your case, you want to do this when you have a value of `base`.

**Reply by jeanlst** [reply]
@uriyyo one thing I have encountered with this approach is that if the user does not put the required "base" field it's raising an `KeyError` exception instead of a `field required` one. From the way it's there I thought the field "base" would be validated before reaching the validator of "top"

**Reply by uriyyo** [reply]
@jeanlst My bad, please use `get` instead of using direct access to value by key:
```py
    @validator("top", always=True)
    def __validate_top__(cls, v: Optional[str], values: Mapping[str, Any]) -> Any:
        if values.get("base") == MyEnum.Manual and v is None:
            raise NoneIsNotAllowedError

        return v
```

**Reply by jeanlst** [reply]
Thanks @uriyyo  that's what I changed it to. However my question still stays, shouldn't it raise `Field required` for the field `base` before it gets to the validator of `top` ?
